src/modules/barrier/barrier.c

```diff
@@ -21,245 +47,357 @@
+        flux_log_error (h, "barrier_lookup_create");
+        return;
     }
+    barrier_update (b, count);
+}
 
-    /* If the count has been reached, terminate the barrier;
-     * o/w set timer to pass count upstream and zero it here.
-     */
-    b->count += count;
-    if (b->count == b->nprocs) {
-        if (exit_event_send (ctx->h, b->name, 0) < 0)
+/* Handle client request to enter barrier.
+ * Response is normally deferred until barrier is complete.
+ */
+static void enter_request_cb (flux_t *h, flux_msg_handler_t *mh,
+                              const flux_msg_t *msg, void *arg)
+{
+    struct barrier_ctx *ctx = arg;
+    struct barrier *b;
+    char *sender = NULL;
+    const char *name;
+    int nprocs;
+    uint32_t owner;
+
+    if (flux_request_unpack (msg,
+                             NULL,
+                             "{s:s s:i !}",
+                             "name", &name,
+                             "nprocs", &nprocs) < 0)
+        goto error;
+    if (flux_msg_get_route_first (msg, &sender) < 0)
+        goto error;
+    if (flux_msg_get_userid (msg, &owner) < 0)
+        goto error;
+    if (!(b = barrier_lookup_create (ctx, name, nprocs, owner)))
+        goto error;
+
+    if (barrier_add_client (b, sender, msg) < 0) {
+        flux_log (ctx->h, LOG_ERR, "abort %s due to double entry by %s",
+                  name, sender);
+        if (exit_event_send (ctx->h, b->name, b->owner, ECONNABORTED) < 0)
             flux_log_error (ctx->h, "exit_event_send");
-    } else if (ctx->rank > 0 && !ctx->timer_armed) {
-        flux_timer_watcher_reset (ctx->timer, barrier_reduction_timeout_sec, 0.);
-        flux_watcher_start (ctx->timer);
-        ctx->timer_armed = true;
+        errno = EEXIST;
+        goto error;
     }
-done:
-    if (sender)
-        free (sender);
+    if (barrier_update (b, 1) < 0)
+        goto error;
+    free (sender);
+    return;
+error:
+    flux_respond_error (ctx->h, msg, errno, NULL);
+    free (sender);
 }
 
 /* Upon client disconnect, abort any pending barriers it was
  * participating in.
  */
-
 static void disconnect_request_cb (flux_t *h, flux_msg_handler_t *mh,
                                    const flux_msg_t *msg, void *arg)
 {
-    barrier_ctx_t *ctx = arg;
+    struct barrier_ctx *ctx = arg;
     char *sender;
     const char *key;
-    barrier_t *b;
+    struct barrier *b;
+    uint32_t userid;
 
-    if (flux_msg_get_route_first (msg, &sender) < 0)
+    if (flux_msg_get_userid (msg, &userid) < 0) {
+        flux_log_error (h, "cannot determine barrier.disconnect userid");
         return;
+    }
+    if (flux_msg_get_route_first (msg, &sender) < 0) {
+        flux_log_error (h, "cannot determine barrier.disconnect sender");
+        return;
+    }
     FOREACH_ZHASH (ctx->barriers, key, b) {
         if (zhash_lookup (b->clients, sender)) {
-            if (exit_event_send (h, b->name, ECONNABORTED) < 0)
+            if (b->owner != userid) {
```

Comment:
Just noticed this: does the disconnect handler really need to check that the userid of the `barrier.disconnect` msg matches the disconnecting barrier "owner" userid? Doesn't the api module send all disconnect messages, and thus disconnect msgs will have userid of the instance owner?

I'm not sure if this is a valid test but ctrl-c out of a barrier with as a non-owner (the overloaded `owner` here might get confusing) does seem to fail, leaving the barrier "active":
```console
$ FLUX_HANDLE_ROLEMASK=0x02 FLUX_HANDLE_USERID=1000 t/barrier/tbarrier --nprocs=2 guest-test 
^C
2019-07-08T22:23:57.853232Z barrier.err[0]: client userid mismatch 6885 != 1000
```

Reply:
Good catch!  Yeah, it should always allow the owner role, but if that's not set, it still probably should check the userid since a disconnect message is not special and could be sent by anyone.